Create a custom standoff mapping

Hi,
I try to create a standoff mapping for our project following the steps of this [doc](https://docs.knora.org/paradox/03-apis/api-v1/xml-to-standoff-mapping.html#creating-a-custom-mapping). 
When I send my [data](https://github.com/dhlab-basel/Knora/files/2122849/data.xml.zip) with a bulk import : 
```
curl -u lumieres@unil.ch:test -X POST -d @data.xml http://localhost:3333/v1/resources/xmlimport/http%3A%2F%2Frdfh.ch%2Fprojects%2F0113
```
I get this error:
```json
{
    "status": 4,
    "error": "net.sf.saxon.s9api.SaxonApiException: org.xml.sax.SAXParseException; lineNumber: 13; columnNumber: 39; Element type \"xsl:template\" must be followed by either attribute specifications, \">\" or \"/>\"."
}
```
1/ do I have to provide a style sheet related to my mapping, is that the problem ?
2/ is the mapping replace or extend the default existing mapping ?

thx for your help

---

My test case is the following:
```html
<text xmlns="">
 <p>test p tag</p>
 <span class="sic" data-corr="Paris">paris</span>
</text>
```
`<p>` tag is already managed by the default mapping of Knora, so in my mapping I just have to map the `<span>` tag of class _sic_ and store the _data-corr_ string value (meaning a data fix): 
```xml
<?xml version="1.0" encoding="UTF-8"?>
<mapping>
    <mappingElement>
        <tag>
            <name>span</name>
            <class>sic</class>
            <namespace>noNamespace</namespace>
            <separatesWords>true</separatesWords>
        </tag>
        <standoffClass>            
            <classIri>http://www.knora.org/ontology/0113/lumieres-lausanne#StandoffEditionTag</classIri>
            <attributes>
                <attribute>
                    <attributeName>data-corr</attributeName>
                    <namespace>noNamespace</namespace>
                    <propertyIri>http://www.knora.org/ontology/0113/lumieres-lausanne#standoffEditionTagHasFix</propertyIri>
                </attribute>
            </attributes>
        </standoffClass>
    </mappingElement>
 </mapping>
```
```json
{
    "project_id": "http://rdfh.ch/projects/0113",
    "label": "LL Standoff mapping",
    "mappingName": "LLStandoffMapping"
}
```
```
curl -u lumieres@unil.ch:test -X POST -F json=@mappingForLL.json -F xml=@mappingForLL.xml  http://localhost:3333/v1/mapping
```
The custom standoff class:
```turtle
lumieres-lausanne:standoffEditionTagHasFix rdf:type owl:DatatypeProperty ;
                  knora-base:subjectClassConstraint lumieres-lausanne:StandoffEditionTag ;
                  knora-base:objectDatatypeConstraint xsd:string .

lumieres-lausanne:StandoffEditionTag rdf:type owl:Class ;
    rdfs:subClassOf knora-base:StandoffTag,
                   [
                      rdf:type owl:Restriction ;
                      owl:onProperty lumieres-lausanne:standoffEditionTagHasFix ;
                      owl:cardinality "1"^^xsd:nonNegativeInteger
                   ] ;
    rdfs:label "Represents an edition fix in a TextValue"@en.     
```

a specific branch of Knora ? no, the develop branch. If you are talking about my project, our git lab is internal only :-|  Our ontology if you want to make some tests: 
[lumieres-lausanne-onto.zip](https://github.com/dhlab-basel/Knora/files/2123146/lumieres-lausanne-onto.zip)

(To reproduce, all my curl commands are in the previous posts above)

Trace: 
```
webapi ----------------------------------------------------------------
webapi Knora API Server started at http://0.0.0.0:3333
webapi ----------------------------------------------------------------
webapi[ERROR] Error on line 13 column 52
webapi[ERROR]   SXXP0003: Error reported by XML parser: Element type "xsl:template" must be followed by
webapi[ERROR]   either attribute specifications, ">" or "/>".
webapi [2018-06-21 12:20:37,713] ERROR - KnoraService(akka://webapi) - Unable to run route /v1/resources/xmlimport/http:%2F%2Frdfh.ch%2Fprojects%2F0113
webapi net.sf.saxon.s9api.SaxonApiException: org.xml.sax.SAXParseException; lineNumber: 13; columnNumber: 52; Element type "xsl:template" must be followed by either attribute specifications, ">" or "/>".
webapi 	at net.sf.saxon.s9api.XsltCompiler.compile(XsltCompiler.java:862)
webapi 	at org.knora.webapi.util.standoff.XMLToStandoffUtil.xml2TextWithStandoff(XMLToStandoffUtil.scala:410)
webapi 	at org.knora.webapi.util.standoff.StandoffTagUtilV2$.convertXMLtoStandoffTagV2(StandoffTagUtilV2.scala:250)
webapi 	at org.knora.webapi.routing.RouteUtilV1$.$anonfun$convertXMLtoStandoffTagV1$2(RouteUtilV1.scala:217)
webapi 	at scala.util.Success.$anonfun$map$1(Try.scala:251)
webapi 	at scala.util.Success.map(Try.scala:209)
webapi 	at scala.concurrent.Future.$anonfun$map$1(Future.scala:289)
webapi 	at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:29)
webapi 	at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:29)
webapi 	at scala.concurrent.impl.CallbackRunnable.run_aroundBody0(Promise.scala:60)
webapi 	at scala.concurrent.impl.CallbackRunnable$AjcClosure1.run(Promise.scala:1)
webapi 	at org.aspectj.runtime.reflect.JoinPointImpl.proceed(JoinPointImpl.java:149)
webapi 	at kamon.scala.instrumentation.FutureInstrumentation.$anonfun$aroundExecution$1(FutureInstrumentation.scala:46)
webapi 	at kamon.Kamon$.withContext(Kamon.scala:120)
webapi 	at kamon.scala.instrumentation.FutureInstrumentation.aroundExecution(FutureInstrumentation.scala:46)
webapi 	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:59)
webapi 	at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:55)
webapi 	at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:91)
webapi 	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
webapi 	at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:81)
webapi 	at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:91)
webapi 	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:40)
webapi 	at kamon.executors.Executors$InstrumentedExecutorService$$anon$7.run(Executors.scala:270)
webapi 	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:43)
webapi 	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
webapi 	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
webapi 	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
webapi 	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
webapi Caused by: net.sf.saxon.trans.XPathException: org.xml.sax.SAXParseException; lineNumber: 13; columnNumber: 52; Element type "xsl:template" must be followed by either attribute specifications, ">" or "/>".
webapi 	at net.sf.saxon.event.Sender.sendSAXSource(Sender.java:471)
webapi 	at net.sf.saxon.event.Sender.send(Sender.java:179)
webapi 	at net.sf.saxon.style.StylesheetModule.loadStylesheet(StylesheetModule.java:233)
webapi 	at net.sf.saxon.style.Compilation.compileSingletonPackage(Compilation.java:101)
webapi 	at net.sf.saxon.s9api.XsltCompiler.compile(XsltCompiler.java:859)
webapi 	... 27 common frames omitted
webapi Caused by: org.xml.sax.SAXParseException: Element type "xsl:template" must be followed by either attribute specifications, ">" or "/>".
webapi 	at org.apache.xerces.util.ErrorHandlerWrapper.createSAXParseException(Unknown Source)
webapi 	at org.apache.xerces.util.ErrorHandlerWrapper.fatalError(Unknown Source)
webapi 	at org.apache.xerces.impl.XMLErrorReporter.reportError(Unknown Source)
webapi 	at org.apache.xerces.impl.XMLErrorReporter.reportError(Unknown Source)
webapi 	at org.apache.xerces.impl.XMLErrorReporter.reportError(Unknown Source)
webapi 	at org.apache.xerces.impl.XMLScanner.reportFatalError(Unknown Source)
webapi 	at org.apache.xerces.impl.XMLNSDocumentScannerImpl.scanStartElement(Unknown Source)
webapi 	at org.apache.xerces.impl.XMLDocumentFragmentScannerImpl$FragmentContentDispatcher.dispatch(Unknown Source)
webapi 	at org.apache.xerces.impl.XMLDocumentFragmentScannerImpl.scanDocument(Unknown Source)
webapi 	at org.apache.xerces.parsers.XML11Configuration.parse(Unknown Source)
webapi 	at org.apache.xerces.parsers.XML11Configuration.parse(Unknown Source)
webapi 	at org.apache.xerces.parsers.XMLParser.parse(Unknown Source)
webapi 	at org.apache.xerces.parsers.AbstractSAXParser.parse(Unknown Source)
webapi 	at org.apache.xerces.jaxp.SAXParserImpl$JAXPSAXParser.parse(Unknown Source)
webapi 	at net.sf.saxon.event.Sender.sendSAXSource(Sender.java:451)
webapi 	... 31 common frames omitted

```

--- a/webapi/src/main/scala/org/knora/webapi/util/standoff/XMLToStandoffUtil.scala
+++ b/webapi/src/main/scala/org/knora/webapi/util/standoff/XMLToStandoffUtil.scala
@@ -220,7 +220,7 @@ case class StandoffDiffDelete(baseStartPosition: Int,
 case class XMLTagSeparatorRequired(maybeNamespace: Option[String], tagname: String, maybeClassname: Option[String]) {
 
     // generate an XPath expression to match this element
-    def toXPath = {
+    def toXPath: String = {
 
         val prefix: String = maybeNamespace match {
             case Some(namespace) => namespace + ":"
@@ -228,7 +228,7 @@ case class XMLTagSeparatorRequired(maybeNamespace: Option[String], tagname: Stri
         }
 
         val classSelector: String = maybeClassname match {
-            case Some(classSel) => s"""[@class=\"$classSel\"]"""
+            case Some(classSel) => s"""[@class='$classSel']""" // use single quotes because the xpath expression is wrapped in double quotes
             case None => ""
         }
 

--- a/webapi/src/test/scala/org/knora/webapi/e2e/v1/StandoffV1R2RSpec.scala
+++ b/webapi/src/test/scala/org/knora/webapi/e2e/v1/StandoffV1R2RSpec.scala
@@ -951,7 +951,7 @@ class StandoffV1R2RSpec extends R2RSpec {
                 assert(status == StatusCodes.OK, "creation of a TextValue from XML returned a non successful HTTP status code: " + responseAs[String])
 
                 thirdTextValueIri.set(ResponseUtils.getStringMemberFromResponse(response, "id"))
-                
+
             }
 
         }
@@ -1190,6 +1190,104 @@ class StandoffV1R2RSpec extends R2RSpec {
 
 
             }
+
+        }
+
+        "create a mapping containing an element with a class that separates words" in {
+
+            val mapping = """<?xml version="1.0" encoding="UTF-8"?>
+                <mapping xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../../../src/main/resources/mappingXMLToStandoff.xsd">
+                    <mappingElement>
+                        <tag><name>text</name>
+                            <class>noClass</class>
+                            <namespace>noNamespace</namespace>
+                            <separatesWords>false</separatesWords></tag>
+                        <standoffClass>
+                            <classIri>http://www.knora.org/ontology/standoff#StandoffRootTag</classIri>
+                            <attributes>
+                                <attribute>
+                                    <attributeName>documentType</attributeName>
+                                    <namespace>noNamespace</namespace>
+                                    <propertyIri>http://www.knora.org/ontology/standoff#standoffRootTagHasDocumentType</propertyIri>
+                                </attribute>
+                            </attributes>
+                        </standoffClass>
+                    </mappingElement>
+
+                    <mappingElement>
+                        <tag>
+                            <name>div</name>
+                            <class>paragraph</class>
+                            <namespace>noNamespace</namespace>
+                            <separatesWords>true</separatesWords>
+                        </tag>
+                        <standoffClass>
+                            <classIri>http://www.knora.org/ontology/standoff#StandoffParagraphTag</classIri>
+                        </standoffClass>
+                    </mappingElement>
+                </mapping>""".stripMargin
+
+            val params =
+                s"""
+                   |{
+                   |  "project_id": "$ANYTHING_PROJECT_IRI",
+                   |  "label": "mapping for elements separating words",
+                   |  "mappingName": "MappingSeparatingWords"
+                   |}
+             """.stripMargin
+
+
+
+            val formDataMapping = Multipart.FormData(
+                Multipart.FormData.BodyPart(
+                    "json",
+                    HttpEntity(ContentTypes.`application/json`, params)
+                ),
+                Multipart.FormData.BodyPart(
+                    "xml",
+                    HttpEntity(ContentTypes.`text/xml(UTF-8)`, mapping),
+                    Map("filename" -> "mapping.xml")
+                )
+            )
+
+            // send mapping xml to route
+            Post("/v1/mapping", formDataMapping) ~> addCredentials(BasicHttpCredentials(anythingUserEmail, password)) ~> standoffPath ~> check {
+
+                assert(status == StatusCodes.OK)
+
+            }
+
+        }
+
+        "create a TextValue from a XML using an element with a class that separates words" in {
+
+            val xmlToSend =
+                """<?xml version="1.0" encoding="UTF-8"?>
+                <text documentType="html">
+                    <div class="paragraph">
+                        This an element that has a class and it separates words.
+                    </div>
+                </text>""".stripMargin
+
+            val newValueParams =
+                s"""
+                    {
+                      "project_id": "http://rdfh.ch/projects/0001",
+                      "res_id": "http://rdfh.ch/0001/a-thing",
+                      "prop": "http://www.knora.org/ontology/0001/anything#hasText",
+                      "richtext_value": {
+                            "xml": ${JsString(xmlToSend)},
+                            "mapping_id": "$ANYTHING_PROJECT_IRI/mappings/MappingSeparatingWords"
+                      }
+                    }""".stripMargin
+
+            // create standoff from XML
+            Post("/v1/values", HttpEntity(ContentTypes.`application/json`, newValueParams)) ~> addCredentials(BasicHttpCredentials(anythingUserEmail, password)) ~> valuesPath ~> check {
+
+                assert(status == StatusCodes.OK, "creation of a TextValue from XML returned a non successful HTTP status code: " + responseAs[String])
+
+            }
+
 
         }
 

--- a/webapi/src/test/scala/org/knora/webapi/util/standoff/XMLToStandoffUtilSpec.scala
+++ b/webapi/src/test/scala/org/knora/webapi/util/standoff/XMLToStandoffUtilSpec.scala
@@ -498,6 +498,35 @@ class XMLToStandoffUtilSpec extends CoreSpec {
 
 
         }
+
+        "convert an XML document containing elements with classes to a TextWithStandoff and check that information separator two has been inserted in the string" in {
+
+            val testXML =
+                """<?xml version="1.0" encoding="UTF-8"?>
+                  |   <text>
+                  |      <text documentType="html">
+                  |                    <div class="paragraph">
+                  |                        This an element that has a class and it separates words.
+                  |                    </div>
+                  |                </text>
+                  |   </text>
+                  |
+                """.stripMargin
+
+
+            val standoffUtil = new XMLToStandoffUtil()
+
+            // after every <div class="paragraph">, information separator two should be inserted
+            val textWithStandoff: TextWithStandoff = standoffUtil.xml2TextWithStandoff(testXML,
+                tagsWithSeparator = List(XMLTagSeparatorRequired(maybeNamespace = None, tagname = "div", maybeClassname = Some("paragraph"))),
+                log = log)
+
+            // make sure that there are as many information separator two as there are paragraphs (there are three paragraphs)
+            assert(StringFormatter.INFORMATION_SEPARATOR_TWO.toString.r.findAllIn(textWithStandoff.text).length == 1)
+
+
+        }
+
     }
 }
 

